packages/lib/src/vis/line/DataCurve.tsx

```diff
@@ -54,15 +61,106 @@ function DataCurve(props: Props) {
     ignoreValue,
   } = props;
 
-  const [dataGeometry] = useState(() => new BufferGeometry());
-  const points = useCanvasPoints(abscissas, ordinates, errors, ignoreValue);
+  const { length } = ordinates;
+  const hasErrors = !!errors;
+  const { abscissaScale, ordinateScale } = useVisCanvasContext();
   const invalidate = useThree((state) => state.invalidate);
 
+  const valueToPosition = useValueToPosition(
+    abscissas,
+    abscissaScale,
+    ordinateScale,
+    ignoreValue
+  );
+
+  const valueToErrorPositions = useValueToErrorPositions(errors, ordinateScale);
+
+  const dataGeometry = useMemo(() => {
+    const geometry = new BufferGeometry();
+    geometry.setAttribute('position', createBufferAttr(length, 3));
+    return geometry;
+  }, [length]);
+
+  const errorGeometries = useMemo(() => {
+    if (!hasErrors) {
+      return undefined;
+    }
+
+    const geometries = {
+      caps: new BufferGeometry(),
+      bars: new BufferGeometry(),
+    };
+
+    geometries.caps.setAttribute('position', createBufferAttr(length * 2, 3));
+    geometries.bars.setAttribute('position', createBufferAttr(length * 2, 3));
+
+    return geometries;
+  }, [hasErrors, length]);
+
+  // eslint-disable-next-line sonarjs/cognitive-complexity
   useLayoutEffect(() => {
-    dataGeometry.setFromPoints(points.data);
+    const dataPosition = dataGeometry.attributes.position;
+    const errorPositions = errorGeometries && {
+      caps: errorGeometries.caps.attributes.position,
+      bars: errorGeometries.bars.attributes.position,
+    };
+
+    ordinates.forEach((value, index) => {
+      const pos = valueToPosition(value, index);
+
+      if (pos) {
+        dataPosition.setXYZ(index, pos[0], pos[1], 0);
+      } else {
+        dataPosition.setXYZ(index, 0, 0, CAMERA_FAR);
+      }
+
+      if (!errorPositions) {
+        return;
+      }
+
+      const { topCap, bottomCap, bar } = valueToErrorPositions(
+        value,
+        index,
+        pos
+      );
+
+      if (topCap) {
+        errorPositions.caps.setXYZ(index * 2 + 1, topCap[0], topCap[1], 0);
+      } else {
+        errorPositions.caps.setXYZ(index * 2 + 1, 0, 0, CAMERA_FAR);
```

Comment:
Again, all the `else` blocks will be hidden away in a method of the future subclass of `BufferGeometry`.